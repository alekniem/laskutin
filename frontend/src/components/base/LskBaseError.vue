<template>
    <div class="red--text font-weight-bold">
        <slot />
    </div>
</template>

<script>
export default {
    name: "LskBaseError",
};
</script>
