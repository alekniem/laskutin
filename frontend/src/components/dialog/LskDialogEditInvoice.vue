<template>
    <v-dialog max-width="700" persistent v-model="dialogVisible">
        <ValidationObserver #default="{ handleSubmit, invalid }" ref="observer">
            <v-form>
                <v-card>
                    <v-card-title :id="cid + 'Title'">
                        {{ title }}
                    </v-card-title>
                    <v-card-text>
                        <v-row dense>
                            <v-col>
                                <LskInputDate
                                    :cid="cid + 'InvoiceDateInput'"
                                    label="Invoice date"
                                    vid="invoice_date"
                                    v-model="inputData.invoice_date"
                                />
                            </v-col>
                            <v-col>
                                <LskInputDate
                                    :cid="cid + 'DueDateInput'"
                                    label="Due date"
                                    vid="due_date"
                                    v-model="inputData.due_date"
                                />
                            </v-col>
                        </v-row>
                        <v-row dense>
                            <v-col>
                                <LskInputText
                                    :cid="cid + 'InvoiceNumberInput'"
                                    label="Invoice number"
                                    vid="invoice_number"
                                    v-model="inputData.invoice_number"
                                />
                            </v-col>
                            <v-col>
                                <LskInputText
                                    :cid="cid + 'ReferenceNumberInput'"
                                    label="Reference number"
                                    vid="reference_number"
                                    v-model="inputData.reference_number"
                                />
                            </v-col>
                        </v-row>
                        <template v-if="otherErrors">
                            <LskBaseError
                                v-for="(error, key) in otherErrors"
                                :key="key"
                            >
                                {{ error | trans }}
                            </LskBaseError>
                        </template>
                    </v-card-text>
                    <v-card-actions>
                        <v-spacer />
                        <LskInputButton
                            @click.native="processCancel"
                            :cid="cid + 'CancelButton'"
                            buttonText="Cancel"
                        />
                        <LskInputButton
                            @click.native="handleSubmit(processSave)"
                            :cid="cid + 'SaveButton'"
                            :disabled="invalid || saveButtonDisabled"
                            buttonText="Save"
                        />
                    </v-card-actions>
                </v-card>
            </v-form>
        </ValidationObserver>
    </v-dialog>
</template>

<script>
import LskDialogAddOrEdit from "@/mixins/LskDialogAddOrEdit";

export default {
    name: "LskDialogEditInvoice",

    mixins: [LskDialogAddOrEdit],

    props: {
        cid: { type: String, required: true },
    },

    data() {
        return {
            fields: [
                "invoice_date",
                "due_date",
                "invoice_number",
                "reference_number",
            ],
        };
    },

    computed: {
        apiActionName() {
            return "invoiceUpdate";
        },

        title() {
            return this.trans("Edit invoice") + " (" + this.id + ")";
        },
    },
};
</script>
