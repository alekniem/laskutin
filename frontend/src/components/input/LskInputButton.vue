<template>
    <v-btn :color="color" :disabled="disabled" :id="cid" depressed small>
        {{ buttonText | trans }}
    </v-btn>
</template>

<script>
export default {
    name: "LskInputButton",

    props: {
        buttonText: { type: String, required: true },
        cid: { type: String, required: true },
        color: { type: String, required: false, default: "primary" },
        disabled: { type: Boolean, required: false, default: false },
    },
};
</script>
