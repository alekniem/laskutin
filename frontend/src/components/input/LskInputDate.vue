<template>
    <ValidationProvider
        :customMessages="{ regex: trans('Invalid date') }"
        :rules="{ regex: /^\d{1,2}\.\d{1,2}\.\d{4}$/ }"
        :vid="vid"
        ref="provider"
        tag="div"
        #default="{ errors, valid }"
    >
        <v-text-field
            :error-messages="errors"
            :id="cid"
            :label="trans(label)"
            :success="valid"
            v-model="textFieldValue"
        />
    </ValidationProvider>
</template>

<script>
import { ValidationProvider } from "vee-validate";

export default {
    name: "LskInputDate",

    components: {
        ValidationProvider,
    },

    props: {
        cid: { type: String, required: true },
        label: { type: String, required: true },
        value: { type: String, required: true },
        vid: { type: String, required: true },
    },

    data() {
        return {
            textFieldValue: null,
        };
    },

    watch: {
        textFieldValue() {
            if (/^\d{1,2}\.\d{1,2}\.\d{4}$/.test(this.textFieldValue)) {
                let parts = this.textFieldValue.split(".");
                let date = new Date(parts[2], parts[1] - 1, parts[0]);
                let result = this.formatDate(date, "YYYY-MM-DD");
                this.$emit("input", result);
            }
        },
    },

    mounted() {
        if (/^\d{4}-\d{2}-\d{2}$/.test(this.value)) {
            this.textFieldValue = this.formatDate(this.value, "D.M.YYYY");
        } else {
            this.textFieldValue = this.value;
        }
    },
};
</script>
