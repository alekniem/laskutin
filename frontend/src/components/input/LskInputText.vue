<template>
    <ValidationProvider
        :customMessages="{ required: trans('Required') }"
        :vid="vid"
        ref="provider"
        rules="required"
        tag="div"
        #default="{ errors, valid }"
    >
        <v-text-field
            @change="$emit('input', $event)"
            :error-messages="errors"
            :id="cid"
            :label="trans(label)"
            :success="valid"
            :value="value"
        />
    </ValidationProvider>
</template>

<script>
import { ValidationProvider } from "vee-validate";

export default {
    name: "LskInputText",

    components: {
        ValidationProvider,
    },

    props: {
        cid: { type: String, required: true },
        label: { type: String, required: true },
        value: { type: [String, Number], required: true },
        vid: { type: String, required: true },
    },
};
</script>
