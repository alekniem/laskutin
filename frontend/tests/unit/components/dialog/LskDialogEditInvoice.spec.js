import flushPromises from "flush-promises";
import LskBaseError from "@/components/base/LskBaseError";
import LskDialogEditInvoice from "@/components/dialog/LskDialogEditInvoice";
import LskInputButton from "@/components/input/LskInputButton";
import LskInputDate from "@/components/input/LskInputDate";
import LskInputText from "@/components/input/LskInputText";
import { createWrapper } from "UT/lsk_test_helper";

const stubs = {
    LskBaseError,
    LskInputButton,
    LskInputDate,
    LskInputText,
};

describe("LskDialogEditInvoice", () => {
    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogEditInvoice, {
            propsData: {
                cid: "TEST_CID",
                id: 123,
                showDialog: false,
            },
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog not visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogEditInvoice, {
            propsData: {
                cid: "TEST_CID",
                id: 123,
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogEditInvoice, {
            propsData: {
                cid: "TEST_CID",
                id: 123,
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();

        const invoiceDateInput = wrapper.find("#TEST_CIDInvoiceDateInput");
        invoiceDateInput.element.value = "INVALID INVOICE DATE";
        invoiceDateInput.trigger("input");
        const dueDateInput = wrapper.find("#TEST_CIDDueDateInput");
        dueDateInput.element.value = "INVALID DUE DATE";
        dueDateInput.trigger("input");
        await flushPromises();

        wrapper.vm._data.inputData.invoice_number = "";
        wrapper.vm._data.inputData.reference_number = "";
        wrapper.vm._data.otherErrors = ["ERROR ONE", "ERROR TWO"];
        await flushPromises();

        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("error messages");
    });

    it("should have correct title", async () => {
        const wrapper = createWrapper(LskDialogEditInvoice, {
            propsData: {
                cid: "TEST_CID",
                id: 123,
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();
        expect(wrapper.find(".v-card__title").text()).toBe("Edit invoice (123)");
    });
});
