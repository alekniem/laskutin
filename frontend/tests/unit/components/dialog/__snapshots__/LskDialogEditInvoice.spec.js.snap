// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LskDialogEditInvoice should match snapshot: dialog not visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
</div>
`;

exports[`LskDialogEditInvoice should match snapshot: dialog visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div class="v-card v-sheet theme--light">
              <div id="TEST_CIDTitle"
                   class="v-card__title"
              >
                Edit invoice (123)
              </div>
              <div class="v-card__text">
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted success--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDInvoiceDateInput"
                                     class="v-label theme--light success--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Invoice date
                              </label>
                              <input id="TEST_CIDInvoiceDateInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light success--text">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted success--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDDueDateInput"
                                     class="v-label theme--light success--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Due date
                              </label>
                              <input id="TEST_CIDDueDateInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light success--text">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input theme--light v-text-field v-text-field--is-booted">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDInvoiceNumberInput"
                                     class="v-label theme--light"
                                     style="left: 0px; position: absolute;"
                              >
                                Invoice number
                              </label>
                              <input id="TEST_CIDInvoiceNumberInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input theme--light v-text-field v-text-field--is-booted">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDReferenceNumberInput"
                                     class="v-label theme--light"
                                     style="left: 0px; position: absolute;"
                              >
                                Reference number
                              </label>
                              <input id="TEST_CIDReferenceNumberInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div class="v-card__actions">
                <div class="spacer">
                </div>
                <button type="button"
                        class="v-btn v-btn--has-bg theme--light v-size--small primary"
                        id="TEST_CIDCancelButton"
                >
                  <span class="v-btn__content">
                    Cancel
                  </span>
                </button>
                <button type="button"
                        disabled="disabled"
                        class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                        id="TEST_CIDSaveButton"
                >
                  <span class="v-btn__content">
                    Save
                  </span>
                </button>
              </div>
            </div>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`LskDialogEditInvoice should match snapshot: error messages 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div class="v-card v-sheet theme--light">
              <div id="TEST_CIDTitle"
                   class="v-card__title"
              >
                Edit invoice (123)
              </div>
              <div class="v-card__text">
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted error--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDInvoiceDateInput"
                                     class="v-label v-label--active theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Invoice date
                              </label>
                              <input id="TEST_CIDInvoiceDateInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Invalid date
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--is-booted error--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDDueDateInput"
                                     class="v-label v-label--active theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Due date
                              </label>
                              <input id="TEST_CIDDueDateInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Invalid date
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted error--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDInvoiceNumberInput"
                                     class="v-label theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Invoice number
                              </label>
                              <input id="TEST_CIDInvoiceNumberInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Required
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted error--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDReferenceNumberInput"
                                     class="v-label theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Reference number
                              </label>
                              <input id="TEST_CIDReferenceNumberInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Required
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="red--text font-weight-bold">
                  ERROR ONE
                </div>
                <div class="red--text font-weight-bold">
                  ERROR TWO
                </div>
              </div>
              <div class="v-card__actions">
                <div class="spacer">
                </div>
                <button type="button"
                        class="v-btn v-btn--has-bg theme--light v-size--small primary"
                        id="TEST_CIDCancelButton"
                >
                  <span class="v-btn__content">
                    Cancel
                  </span>
                </button>
                <button type="button"
                        disabled="disabled"
                        class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                        id="TEST_CIDSaveButton"
                >
                  <span class="v-btn__content">
                    Save
                  </span>
                </button>
              </div>
            </div>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;
